### DISCUSSION TITLE
Weird union query compile

### DISCUSSION THREAD

**Discussion by jossefaz** [sqlalchemy/sqlalchemy#7804] [Usage Questions]
I have two simple queries. I want the union of them

First query 

```python
specific_query = session.query(kv).join(kvi).filter(kv.org_id ==2)
```
In SQL (I just print this `specific_query` to the console) :
```sql
SELECT kvi.id                              AS kvi_id,
       kvi.vid                             AS kvi_vid,
       kvi.interpretation                  AS kvi_interpretation,
       kv.id                               AS kv_id,
       kv.org_id                           AS kv_org_id,
       kv.position                         AS kv_position
FROM kv INNER JOIN kvi ON kv.id = kvi.vid
WHERE kv.org_id = 2
```

Second query

```python
v1 = aliased(Variant)
v2 = aliased(Variant)
sub_exists = session.query(1).filter(and_(v2.position == v1.position, v2.org_id == 2)).exists()
exclude_query = session.query(v1).join(kvi).filter(~sub_exists, v1.org_id.in_([3,4]))
```
In SQL 
```sql
SELECT kvi.id                              AS kvi_id,
       kvi.vid                             AS kvi_vid,
       kvi.interpretation                  AS kvi_interpretation,
       kv.id                               AS kv_id,
       kv.org_id                           AS kv_org_id,
       kv.position                         AS kv_position,
FROM kv AS kv_1 INNER JOIN kvi ON kv_1.id = kvi.vid
WHERE NOT (EXISTS(SELECT 1
                  FROM kv AS kv_2
                  WHERE kv_2.position = kv_1.position
                    AND kv_2.org_id = 2))
AND kv_1.org_id IN (3,4)
```
Everything seems good. And if I execute those queries separately it works fine. 

My goal is to get the union of those two

```python
my_query = specific_query.union(exclude_query)
```

**Expected output :** 

```sql
SELECT kvi.id                              AS kvi_id,
       kvi.vid                             AS kvi_vid,
       kvi.interpretation                  AS kvi_interpretation,
       kv.id                               AS kv_id,
       kv.org_id                           AS kv_org_id,
       kv.position                         AS kv_position
FROM kv INNER JOIN kvi ON kv.id = kvi.variant_id
WHERE kv.org_id = 2
UNION
SELECT kvi.id                              AS kvi_id,
       kvi.vid                             AS kvi_vid,
       kvi.interpretation                  AS kvi_interpretation,
       kv.id                               AS kv_id,
       kv.org_id                           AS kv_org_id,
       kv.position                         AS kv_position,
FROM kv AS kv_1 INNER JOIN kvi ON kv_1.id = kvi.vid
WHERE NOT (EXISTS(SELECT 1
                  FROM kv AS kv_2
                  WHERE kv_2.position = kv_1.position
                    AND kv_2.org_id = 2))
  AND kv_1.org_id IN (3,4)
```
When I execute this **_manually_** it works perfectly
However, when I try with the code, it first compiles to this (strange) sql query : 

**Actual output**
```Sql
SELECT kvi.id             AS kvi_id,
       kvi.vid            AS kvi_vid,
       kvi.interpretation AS kvi_interpretation,
       anon_1.kv_id       AS anon_1_kv_id,
       anon_1.kv_org_id   AS anon_1_kv_org_id,
       anon_1.kv_position AS anon_1_kv_position,
FROM kvi,
     (SELECT kv.id       AS kv_id,
             kv.org_id   AS kv_org_id,
             kv.position AS kv_position
      FROM kv
               INNER JOIN kvi ON kv.id = kvi.vid
      WHERE kv.org_id = 2
      UNION
      SELECT kv.id       AS kv_id,
             kv.org_id   AS kv_org_id,
             kv.position AS kv_position
      FROM kv AS kv_1
               INNER JOIN kvi
                          ON kv.id = kvi.vid
      WHERE NOT (EXISTS(SELECT 1
                        FROM kv AS kv_2
                        WHERE kv_2.position = kv_1.position
                        AND kv_2.org_id = 2))
        AND kv_1.org_id IN (3, 4)) AS anon_1
```

I want to add a order by statement to this query.
```python
my_query = specific_query.union(exclude_query).order_by(kv.creation_date.asc())
```
In the simple "manual" build union : order by works perfectly (I simply add the `ORDER BY` statement to the SQL)
In the automatic build union (the last query), order by simply not working (It returns me only one row where I expect hundreds)

Help please !!

**Comment by zzzeek** [comment]
query.union() is awful for these use cases which is one of the main reasons I decided to [replace query entirely with select()](https://docs.sqlalchemy.org/en/14/changelog/migration_14.html#orm-query-is-internally-unified-with-select-update-delete-2-0-style-execution-available).  Like, this was the main reason I decided to make the whole 2.0 change.

I haven't read the specifics of your case yet (since it looks like all the old query.union() use cases that I already know work terribly).   Can you try rewriting the union etc. using select(), and then to get an ORM query use the technique documented at https://docs.sqlalchemy.org/en/14/orm/queryguide.html#selecting-entities-from-unions-and-other-set-operations to get ORM entities from it?   I would only be interested in getting this to work using the newer APIs.  thanks!

**Reply by jossefaz** [reply]
Migrating to 2.0 definitely fixed this issue ! 
However : relationships fields now appears as plain fields (and duplicated rows) and not as a list of rows like it was before.

```python
specific_query = select(kv, kvi).join(kvi).where(kv.org_id ==2)
v1 = aliased(Variant)
v2 = aliased(Variant)
sub_exists = select(1).where(and_(v2.position == v1.position, v2.org_id == 2)).exists()
exclude_query = select(v1, kvi).join(kvi).where(~sub_exists, v1.org_id.in_([3,4]))
result = session.execute(specific_query.union(exclude_query)).unique().all()
```
For each `kvi` (the relationship of kv table) I receive a new record.
I was regular to retrieve the "interpretation" field of a kvi like so : 

```python
result = session.execute(specific_query.union(exclude_query)).unique().all()
for kv in result :
    for kvi in kv.kvis:
          # Do some stuff
```
But with the new code, I get an `AttributeError` : 
```python
AttributeError: Could not locate column in row for column 'kvis'
```
This is because it put me all the kvi column at the same level as the parent row
 How can I get rows with a `kvis` field which represents a list of all kvis for this row (as before with old syntax) ?

**Reply by zzzeek** [reply]
it looks like you are selecting two entities:  `select(ent1, ent2)` (though it's hard to tell because none of the code above seems to express what kinds of objects "kv" and "kvi" are).  that will return Row objects with one column per entity.   you'd want to select just one entity, like `select(ent1)`, then use [session.scalars()](https://docs.sqlalchemy.org/en/14/orm/session_api.html?highlight=session%20scalars#sqlalchemy.orm.Session.scalars) to get results as a list.

overall though I dont have enough context to see what everything is and how you are calling it, so would need a complete, self-contained runnable [MCVE](http://stackoverflow.com/help/mcve) to understand what you're doing.

**Reply by jossefaz** [reply]
Basicaly I have two tables : `kv` and `Kvi`

They are related by a foreign key on `kvi` that points to `kv.id`

This is a one to many relationship (for one kv it could be many kvis)

On the `kv` model I decalred a relationship field called "kvis" that points to the `kvi` table, with a lazy load strategy (the default)

With the previous syntax, when I query like this : 

```python
specific_query = session.query(kv).join(kvi).options(joinedload(kv.kvis)).filter(kv.org_id ==2).all()
```
As a result I got a list of kv instances. For each instances I was able to get all its kvis by calling the relationship field : `kv.kvis` which returns list of kvi instances.
This is basic relationship usage.

However, with the new syntaxe, I cannot get this relationship field, and I don't know why 

```python
specific_query = select(kv, kvi).join(kvi).options(joinedload(kv.kvis)).where(kv.org_id ==2)
result = session.execute(specific_query).scalars().all()
```
This returns me a list of named tuples with all the kv's attribute but the `kvis` relationship field is missing on each instances. When I try to call it it give me the AttributeError
```
AttributeError: Could not locate column in row for column 'kvis'
```

**Reply by jossefaz** [reply]
Another issue that occurs with the new syntax is the usage of the order_by :

Previous syntax

 `session.query(kv).join(kvi).options (joinedload(kv.kvis)).order_by(kvi.some_field)`

SQL output

`ORDER BY kvi.some_field`

New syntax
`select(kv).join(kvi).options (joinedload(kv.kvis)).order_by(kvi.some_field)`

SQL output 

`ORDER BY some_field`

The explicit call to the table `kvi` is missing

**Reply by CaselIT** [reply]
> However, with the new syntaxe, I cannot get this relationship field, and I don't know why

Your are selecting two elements from the select: ` select(kv, kvi)`. What you are getting back is a `Row` with two elements.
You should take a look at the tutorial: https://docs.sqlalchemy.org/en/14/tutorial/data_select.html#tutorial-selecting-data

Also looking at your query ` select(kv).join(kvi).options (joinedload(kv.kvis)).order_by(kvi.some_field)` you probably want to use contains_eager instead of joinedload

**Reply by jossefaz** [reply]
When I do select only one entity 
`select(kv).join(kvi).options (contains_eager(kv.kvis)).order_by(kvi.some_field)`

I have two issues :
1) without the order by : I got a list of `kv` instances but without kvis relationship field. So I cannot do `kv.kvis` to get a list of kvi instances

2) with the order_by I got an error of unknown column `some_field` because for some reason the order by method removed the explicit call to the table `kvi` (kvi.some_field) to which it belongs

**Reply by CaselIT** [reply]
Again, please read the documentation...

When you do `execute(select(something))` you get back a list of `Row` with one element. To get back a list of elements you need to use `scalars(...)` like mike suggested, or do `execute(...).scalars()`

> 2. with the order_by I got an error of unknown column some_field because for some reason the order by method removed the explicit call to the table kvi (kvi.some_field) to which it belongs

since you haven't provided any example model to reproduce the issue, it's hard to know what's happening

**Reply by zzzeek** [reply]
we really need a complete example to understand what you're trying to achieve and how you are doing it.

**Comment by jossefaz** [comment]
@zzzeek @CaselIT : Here is a complete example in Jupyter notebook (you can even run it). 
I added few examples of the issues that I currently try to figure out : 

**TL;DR** : ORDER BY fields from a relational table does not work when using it with a UNION statement

https://github.com/jossefaz/sqlalchemy-poc-union-/blob/master/union_scalars.ipynb

**Reply by jossefaz** [reply]
Okay I found this part of the documentation specific to UNION statement
https://docs.sqlalchemy.org/en/14/tutorial/data_select.html#selecting-orm-entities-from-unions

Working from this. Will update soon

**Reply by jossefaz** [reply]
I manage to resolve my issue with ORM instances when unsing the UNION statement based on the documentation above.
```python
stmt = select(kv).from_statement(specific_query.union(exclude_query))
```
However, I still have an issue with the ORDER BY statement. The documentation show an example of using ORDER BY with union, but this is only valid for the main table that we select. I did not find in the documentation a part that explain how to use ORDER BY and UNION on a relational table attribute.

I updated the Jupyter notebook to show the issue
https://github.com/jossefaz/sqlalchemy-poc-union-/blob/master/union_scalars.ipynb

**Reply by CaselIT** [reply]
First of all note that you have a cartesian product in the query.
Also I think you could try using `table_alias.interpretation` for the order by

**Reply by jossefaz** [reply]
`table_alias.interpretation` give me `AttributeError: type object 'Kv' has no attribute 'interpretation'`

**Reply by jossefaz** [reply]
BUT `table_alias.kvis.comparator.mapper.c.interpretation` seems to work

**Reply by CaselIT** [reply]
> `table_alias.interpretation` give me `AttributeError: type object 'Kv' has no attribute 'interpretation'`

`table_alias = aliased(Kvi, my_query.subquery())` is the one you should use. You seem to have used the other one.


btw the "correct" query is the following one 
```py
specific_query = select(Kv, Kvi).join(Kvi).options(contains_eager(Kv.kvis)).filter(Kv.org_id ==2)

v1 = aliased(Kv)
v2 = aliased(Kv)
sub_exists = select(1).filter(and_(v2.position == v1.position, v2.org_id == 2)).exists()
exclude_query = select(Kv, Kvi).join(Kvi).options(contains_eager(Kv.kvis)).filter(~sub_exists, v1.org_id.in_([1,3]))
with Session() as session :
    my_query = exclude_query.union(specific_query)
    table_alias = aliased(Kvi, my_query.subquery())
    order_by_query = select(table_alias).order_by(asc(table_alias.interpretation))
    print(order_by_query.compile(engine))
```

but as mentioned it above a cartesian join, so you may want to check that
